### ISSUE TITLE
Quick feedback on interview assignment on 5/13

### ISSUE THREAD

**Issue by boan-smartsurgerytek** [Winston-Ch/ShapeClassifierPyTorch#1]
Thank you for submitting your assignment—we appreciate the time and effort you put into it.
We would like to seek for the candidate who have ability to solve the problem on reality.
Normally, when we sort the problem of black box approach, how to design the experiments is crucial.

1. Shape_Classifier.ipynb

```
        if condition == 'fixed_length_fixed_rotation':
            size = 60
            rotation_angle = 0
        elif condition == 'fixed_length_random_rotation':
            size = 60
            rotation_angle = random.randint(0, 360)
        elif condition == 'random_length_fixed_rotation':
            size = random.randint(30, 80)
            rotation_angle = 0
        elif condition == 'random_length_random_rotation':
```
You have design these experiment and run them: can we get better understanding when given this control condition?

```
patience = 15
```
You have using the patience factor but this factor is unused.

```
x, y = generate_dataset(num_samples=1000)
train_size = int(0.8 * len(dataset))
val_size = int(0.1 * len(dataset))
test_size = len(dataset) - train_size - val_size
```
If you set the val and test, the cross validation can be shown.

```
Epoch 50/50, Train Loss: 0.6812, Train Accuracy: 69.75%, Val Loss: 0.8664, Val Accuracy: 50.00%
```
So far the problem is not solved yet.

```
Possible improvements include increasing dataset size, using more advanced architectures
```
You can run with more experiments to POC.

2. Bonus.ipynb

a. 
The problem can be addressed using Bayesian probability and is expected to exhibit a nonlinear p-t graph.
You can check the formula, something is wrong.

b.
```
after controlling for gender, we see that the medication is not as effective
```
Which assumption is using in your argument? If confounding factor is considered, what about the argument of mediation?
Show whether these two arguments conflict or not?
->
When we should take conditions on sub group? Is it necessary to take consideration on sub group?

**Comment by Winston-Ch** [comment]
@boan-smartsurgerytek 

According to your first question's answer:

        if condition == 'fixed_length_fixed_rotation':
            size = 60
            rotation_angle = 0
        elif condition == 'fixed_length_random_rotation':
            size = 60
            rotation_angle = random.randint(0, 360)
        elif condition == 'random_length_fixed_rotation':
            size = random.randint(30, 80)
            rotation_angle = 0
        elif condition == 'random_length_random_rotation':

For data generation and preprocessing, we can write a function to generate images of circles, squares, and triangles. Each shape should have random properties (size, position) and be randomly placed within the image frame for dataset diversity.

Experimental Conditions:

        fixed_length_fixed_rotation:
        Tests the model's accuracy with shapes of fixed size and rotation, under simple conditions.

        fixed_length_random_rotation:
        Assesses the model’s ability to recognize shapes with fixed size but random rotation, testing rotation invariance.

        random_length_fixed_rotation:
        Evaluates the model's recognition of shapes with varying sizes but fixed rotation.

        random_length_random_rotation:
        The most challenging condition, testing performance with both random size and rotation changes.


Below is the response to your second question.

        patience = 15

The **patience** parameter isn't being used because the model's validation loss consistently improves or fluctuates, preventing early stopping from being triggered. Early stopping requires no improvement in validation loss for a set number of epochs to halt training. 

I have modified the script to implement early stopping and added more samples to my model and uploaded it again. please feel free to check it out and let me know if you have any questions.

Thanks in advance.
